src/ASM/ASMs3D.C

```diff
@@ -3791,3 +3759,165 @@ bool ASMs3D::addRigidCpl (int lindx, int ldim, int basis,
+
+  // Compute parameter values of the Gauss points over the whole patch
+  for (int d = 0; d < 3; d++) {
+    patch.getGaussPointParameters(mainQ.gpar[d],d,mainQ.ng[d],mainQ.xg[d]);
+    if (reducedQ.xg[0])
+      patch.getGaussPointParameters(reducedQ.gpar[d],d,nRed,mainQ.xg[0]);
+  }
+
+  const int p1 = patch.svol->order(0);
+  const int p2 = patch.svol->order(1);
+  const int p3 = patch.svol->order(1);
+  const int n1 = patch.svol->numCoefs(0);
+  const int n2 = patch.svol->numCoefs(1);
+  const int n3 = patch.svol->numCoefs(1);
+  nel1 = n1 - p1 + 1;
+  nel2 = n2 - p2 + 1;
+  const int nel3 = n3 - p3 + 1;
+
+  nTotal = nel1*nel2*nel3*mainQ.ng[0]*mainQ.ng[1]*mainQ.ng[2];
+  if (reducedQ.xg[0])
+    nTotalRed = nel1*nel2*nel3*mainQ.ng[0]*mainQ.ng[1]*mainQ.ng[1];
+
+  return true;
+}
+
+
+double ASMs3D::BasisFunctionCache::getParam (int dir, size_t gp,
+                                             size_t el, bool reduced) const
+{
+  const Quadrature& q = reduced ? reducedQ : mainQ;
+  return q.gpar[dir](gp+1,el+1);
+}
+
+
+BasisFunctionVals ASMs3D::BasisFunctionCache::calculatePt (size_t el,
+                                                           size_t gp,
+                                                           bool reduced) const
+{
+  PROFILE2("Spline evaluation");
+  std::array<size_t,3> gpIdx = this->gpIndex(gp,reduced);
+  std::array<size_t,3> elIdx = this->elmIndex(gp);
+
+  BasisFunctionVals result;
+  if (nderiv == 1) {
+    Go::BasisDerivs spline;
+#pragma omp critical
+    patch.getBasis(basis)->computeBasis(this->getParam(0,gpIdx[0],elIdx[0],reduced),
+                                        this->getParam(1,gpIdx[1],elIdx[1],reduced),
+                                        this->getParam(2,gpIdx[2],elIdx[2],reduced), spline);
+    SplineUtils::extractBasis(spline,result.N,result.dNdu);
+  } else if (nderiv == 2) {
+    Go::BasisDerivs2 spline;
+#pragma omp critical
+    patch.getBasis(basis)->computeBasis(this->getParam(0,gpIdx[0],elIdx[0],reduced),
+                                        this->getParam(1,gpIdx[1],elIdx[1],reduced),
+                                        this->getParam(2,gpIdx[2],elIdx[2],reduced), spline);
+    SplineUtils::extractBasis(spline,result.N,result.dNdu,result.d2Ndu2);
+  }
+
+  return result;
+}
+
+
+size_t ASMs3D::BasisFunctionCache::index (size_t el, size_t gp, bool reduced) const
+{
+  const Quadrature& q = reduced ? reducedQ : mainQ;
+  std::array<size_t,3> elIdx = this->elmIndex(el);
+  std::array<size_t,3> gpIdx = this->gpIndex(gp,reduced);
+
+  return ((elIdx[2]*q.ng[2]*nel2 + elIdx[1])*q.ng[1]*nel1 + elIdx[0])*q.ng[0]
+          + (gpIdx[2]*q.ng[1]*nel2 + gpIdx[1])*q.ng[0]*nel1 + gpIdx[0];
+}
+
+
+std::array<size_t, 3> ASMs3D::BasisFunctionCache::elmIndex (size_t el) const
+{
+  std::array<size_t, 3> result;
+
+  result[0] = el % nel1;
+  result[1] = (el / nel1) % nel2;
+  result[2] = el / (nel1*nel2);
+
+  return result;
```

Comment:
could shorten this to just `return { el % nel1, (el/nel1) % nel2, el / (nel1*nel2) };` ?

Reply:
sure